### ISSUE TITLE
Please provied a Pytorch Lightning LightningModule API

### ISSUE THREAD

**Issue by pfeatherstone** [ultralytics/ultralytics#2284]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

It would be great if there was an API that used Pytorch Lightning's `LightningModule` and `Trainer` objects for training. This would simplify the code base and could just focus on model definition and loss function. Would be great for researchers too.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by AyushExel** [comment]
We've explored many frameworks for building this tool including lightning but we prefer keeping minimum dependencies. 
Do you have any specific feature requests or problems that you think would be solved by using lightning?

**Comment by AyushExel** [comment]
Thanks for the feedback. We'll try to improve the codebase but some parts like DDP for cli and python are quite messy due to some constraints and we looked at the implementation in lightning which was pretty similar internally in the repo.
We prefer to keep the fine control of the codebase which is why we keep the dependencies to a minimum. If there's a specific part of the codebase that you think is messy and needs more documentation, please let us know and we'll prioritize it.

**Comment by pfeatherstone** [comment]
To be honest, if there is something that lightning doesn't do, which you guys need, I would submit a feature request to lightning. It would likely get noticed and implemented if coming from this repo.

**Comment by glenn-jocher** [comment]
@pfeatherstone thank you for your feedback, we appreciate it. Our aim is to keep the codebase as simple and straightforward as possible while still providing the necessary features and control for advanced users. While we currently do not use PyTorch Lightning's `LightningModule` and `Trainer` objects for YOLOv8, we are open to exploring new and better ways to improve the codebase in the future. In the meantime, if there is anything specific that we can add or improve to the existing codebase, please do not hesitate to let us know. We will also consider submitting feature requests to PyTorch Lightning as necessary. Thank you for your continued support of YOLOv8.

**Comment by pfeatherstone** [comment]
Wow, you guys are super nice!

**Comment by robmarkcole** [comment]
@glenn-jocher can I suggest as a minimum, some documentation on using Lightning and ultralytics?

**Comment by glenn-jocher** [comment]
@robmarkcole of course, that would be great! We don't have any Lightning integration though, so I don't know what the documentation would encompass, but if you'd like to take the lead on a new Lightning docs page that focuses on possible integrations or synergies there that would be wonderful. 

I met William Falcon at GTE a few months ago, he's pretty hyperfocused, but the conversation was more sales-y about Lightning than about possible integrations unfortunately.

**Comment by pfeatherstone** [comment]
Looking at ultralytics source code, it looks like you guys have implemented something that approximates lightning, even down to the API. You have hooks, callbacks and other bits and bobs that mirror lightning. May as well just use it. Then all you have to worry about is model definition, TAL and loss

**Comment by pderrenger** [comment]
Thank you for your observation. While our implementation does share similarities with PyTorch Lightning, we prefer to maintain fine control over our codebase and minimize dependencies. This approach allows us to tailor the framework specifically to our needs and ensure optimal performance. However, we appreciate your feedback and will consider it as we continue to evolve our codebase. If you have any specific suggestions or areas you think could benefit from further documentation or improvement, please let us know.

**Comment by Borda** [comment]
I am experimenting with one over here... https://github.com/Borda/lightning-YOLOs
